From 5b686560a5422278bb12d30c21a420d4b574c303 Mon Sep 17 00:00:00 2001
From: Hiroaki Nakamura <hnakamur@gmail.com>
Date: Sun, 27 Jul 2025 16:56:39 +0900
Subject: [PATCH 1/2] Add limit_upstream_status field in http request structure

---
 src/http/ngx_http_request.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/http/ngx_http_request.h b/src/http/ngx_http_request.h
index 9407f46ae..83819bbbe 100644
--- a/src/http/ngx_http_request.h
+++ b/src/http/ngx_http_request.h
@@ -571,6 +571,13 @@ struct ngx_http_request_s {
     unsigned                          background:1;
     unsigned                          health_check:1;
 
+    /*
+     * instead of using the context data in
+     * ngx_http_limit_upstream_module
+     * we use the bit fields in the request structure
+     */
+    unsigned                          limit_upstream_status:3;
+
     /* used to parse HTTP headers */
 
     ngx_uint_t                        state;
-- 
2.43.0

